core/src/main/java/hudson/model/AbstractItem.java

```diff
@@ -228,6 +233,102 @@ protected void doSetName(String name) {
         this.name = name;
     }
 
+    /**
+     * @return whether {@link #name} can be modified by a user
+     */
+    public boolean isNameEditable() {
+        return false;
+    }
+
+    /**
+     * Renames this item
+     */
+    @RequirePOST
+    @Restricted(NoExternalUse.class)
+    public void doDoRename2(StaplerRequest req, StaplerResponse rsp) throws IOException {
+        String newName = req.getParameter("newName");
+        newName = newName == null ? null : newName.trim();
+        FormValidation validationError = doCheckNewName(newName);
```

Comment:
So IIUC this will wind up enforcing `checkRename` at the actual moment of renaming, as well as during form validation—good.